xds/src/main/java/io/grpc/xds/LoadBalancerConfigFactory.java

```diff
@@ -0,0 +1,326 @@
+          logger.log(XdsLogLevel.WARNING, "Policy {0} not found in the LB registry, skipping",
+              typedConfig.getTypeUrl());
+          continue;
+        } else {
+          return serviceConfig;
+        }
+      }
+
+      // If we could not find a Policy that we could both convert as well as find a provider for
+      // then we have an invalid LB policy configuration.
+      throw new ResourceInvalidException("Invalid LoadBalancingPolicy: " + loadBalancingPolicy);
+    }
+
+    /**
+     * Converts a ring_hash {@link Any} configuration to service config format.
+     */
+    private static ImmutableMap<String, ?> convertRingHashConfig(Any config)
+        throws InvalidProtocolBufferException, ResourceInvalidException {
+      RingHash ringHash = config.unpack(RingHash.class);
+
+      // The hash function needs to be validated here as it is not exposed in the returned
+      // configuration for later validation.
+      if (RingHash.HashFunction.XX_HASH != ringHash.getHashFunction()) {
+        throw new ResourceInvalidException(
+            "Invalid ring hash function: " + ringHash.getHashFunction());
+      }
+
+      return buildRingHashConfig(
+          ringHash.hasMinimumRingSize() ? ringHash.getMinimumRingSize().getValue() : null,
+          ringHash.hasMaximumRingSize() ? ringHash.getMaximumRingSize().getValue() : null);
+    }
+
+    /**
+     * Converts a wrr_locality {@link Any} configuration to service config format.
+     */
+    private static ImmutableMap<String, ?> convertWrrLocalityConfig(Any config, int recursionDepth)
+        throws InvalidProtocolBufferException, ResourceInvalidException {
+      WrrLocality wrrLocality = config.unpack(WrrLocality.class);
+      return buildWrrLocalityConfig(
+          convertToServiceConfig(wrrLocality.getEndpointPickingPolicy(), recursionDepth + 1));
+    }
+
+    /**
+     * "Converts" a round_robin configuration to service config format.
+     */
+    private static ImmutableMap<String, ?> convertRoundRobinConfig() {
+      return buildRoundRobinConfig();
+    }
+
+    /**
+     * Converts a custom LB config {@link Any} configuration to service config format.
+     */
+    @SuppressWarnings("unchecked")
+    private static ImmutableMap<String, ?> convertCustomConfig(Any config)
+        throws InvalidProtocolBufferException, ResourceInvalidException {
+      TypedStruct configTypedStruct = config.unpack(TypedStruct.class);
+      Object rawJsonConfig = null;
+      try {
+        rawJsonConfig = JsonParser.parse(JsonFormat.printer().print(configTypedStruct.getValue()));
+      } catch (IOException e) {
+        throw new ResourceInvalidException("Unable to parse custom LB config JSON", e);
+      }
+
+      if (!(rawJsonConfig instanceof Map)) {
+        throw new ResourceInvalidException("Custom LB config does not contain a JSON object");
+      }
+
+      return ImmutableMap.of(
+          configTypedStruct.getTypeUrl().replaceFirst(TYPE_URL_PREFIX_REGEX, ""),
```

Comment:
This needs updating to not assume the particular prefix string.

Reply:
Forgot to do that, done.